src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,421 @@
+/-
+Copyright (c) 2022 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Dilations of emetric and metric spaces
+Authors: Hanting Zhang
+-/
+import topology.metric_space.antilipschitz
+import data.fun_like.basic
+
+/-!
+# Dilations
+We define dilations, i.e., maps between emetric spaces that
+satisfy `edist (f x) (f y) = r * edist x y`.
+Defines `ratio f`, which is the ratio of some `f : dilation α β`.
+Note that `ratio f : ℝ≥0`, so we do not exclude the degenerate case of dilations
+which collapse into constant maps. Statements that do need strict dilations should
+say `f : dilation α β` and `hr : ratio f ≠ 0`.
+TODO: Introduce dilation equivs. Refactor the isometry API
+to match the `*_hom_class` API below.
+Since a lot of elementary properties don't require `eq_of_dist_eq_zero`
+we start setting up the theory for `pseudo_emetric_space` and
+we specialize to `pseudo_metric_space` and `metric_space` when needed.
+# Notes
+The type of dilations defined in this file are also referred to as
+"similiarties" or "similitudes" by other authors. The name `dilation` was choosen
+to match the Wikipedia name.
+# References
+- https://en.wikipedia.org/wiki/Dilation_(metric_space)
+- Marcel Berger, Geometry
```

Comment:
```suggestion
# Dilations

We define dilations, i.e., maps between emetric spaces that
satisfy `edist (f x) (f y) = r * edist x y`.

## Main defintions

* `dilation.ratio f`: the value of `r` in the relation above, defaulting to 1 in the case where it is not well-defined.
  Note that `ratio f : ℝ≥0`, so we do not exclude the degenerate case of dilations
  which collapse into constant maps. Statements that do need strict dilations should
  say `f : dilation α β` and `hr : ratio f ≠ 0`.
  
## Implementation notes

The type of dilations defined in this file are also referred to as
"similiarties" or "similitudes" by other authors. The name `dilation` was choosen
to match the Wikipedia name.

Since a lot of elementary properties don't require `eq_of_dist_eq_zero`
we start setting up the theory for `pseudo_emetric_space` and
we specialize to `pseudo_metric_space` and `metric_space` when needed.

TODO: Introduce dilation equivs. Refactor the `isometry` API
to match the `*_hom_class` API below.

## References

- https://en.wikipedia.org/wiki/Dilation_(metric_space)
- Marcel Berger, Geometry
```

Can you add a bibtex entry for "Marcel Berger" to `references.bib`?

Also, this line-wrapping looks all over the place; can you wrap the above to 100 columns?